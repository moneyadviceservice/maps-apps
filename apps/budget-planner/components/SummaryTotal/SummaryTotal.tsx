import { DetailedHTMLProps, HTMLAttributes, ReactNode } from 'react';

import { twMerge } from 'tailwind-merge';

import { Heading } from '@maps-react/common/components/Heading';
import {
  Props as SelectProps,
  Select,
} from '@maps-react/form/components/Select';
import { useTranslation } from '@maps-react/hooks/useTranslation';

export type Item = DetailedHTMLProps<
  HTMLAttributes<HTMLTableRowElement>,
  HTMLTableRowElement
> & {
  label: string;
  value: ReactNode;
  background?: string;
  isEstimate?: boolean;
};

const Item = ({
  label,
  value,
  className,
  background = 'bg-gray-100',
  isEstimate,
  ...rest
}: Item) => {
  return (
    <tr
      data-testid="summary-total-item-value"
      className={twMerge(
        `flex justify-between print:border-none border w-full h-full rounded-[4px] mb-2 py-5 px-4 ${background} ${
          isEstimate
            ? 'rounded-none border-8 border-y-0 rounded-bl-lg border-l-amber-500 text-black'
            : ''
        }`,
        className,
      )}
      {...rest}
    >
      <th className="text-lg font-normal">{label}</th>
      <td className="text-md" data-testid="item-value">
        {value}
      </td>
    </tr>
  );
};

export type Props = {
  children?: ReactNode;
  onSelectChange?: SelectProps['onChange'];
  selectOptions?: SelectProps['options'];
  items: Item[];
  summary: Item;
  defaultValue?: number;
  showHeading?: boolean;
  isProgress?: boolean;
};

export default function SummaryTotal({
  children,
  onSelectChange,
  defaultValue,
  selectOptions,
  items,
  summary,
  showHeading,
  isProgress,
}: Readonly<Props>) {
  const { z } = useTranslation();
  return (
    <div
      className={twMerge(
        `flex flex-col  print:border-none border border-slate-400 rounded-[4px] px-1 pb-4 ${
          !showHeading ? 'pt-3 pb-4' : ''
        }`,
      )}
    >
      {showHeading ? (
        <div className="flex items-center justify-center px-1 py-4 text-center md:flex-col lg:flex-row print:text-left print:p-2">
          <Heading level="h4" component="h3">
            {z({
              en: `${isProgress ? 'Your progress' : 'Summary total'}`,
              cy: `${isProgress ? 'Eich cynnydd' : 'Cyfanswm crynodeb'}`,
            })}
          </Heading>
          {isProgress && (
            <span className="ml-1 text-lg">
              {z({
                en: '(monthly)',
                cy: '(misol)',
              })}
            </span>
          )}
        </div>
      ) : null}
      {selectOptions && (
        <div className="px-6 pt-0 pb-4">
          <Select
            name="divisor"
            hideEmptyItem
            defaultValue={defaultValue ?? 1}
            onChange={onSelectChange}
            options={selectOptions}
          />
        </div>
      )}
      <table className="w-full border-collapse">
        <thead className="sr-only">
          <tr>
            <th className="hidden"></th>
            <th>Amount</th>
          </tr>
        </thead>
        <tbody className="flex flex-col mx-1">
          {items.map((item) => (
            <Item
              key={item.label}
              label={item.label}
              value={item.value}
              className={item.className}
              isEstimate={item.isEstimate}
            />
          ))}
          <Item
            label={summary.label}
            value={summary.value}
            background={summary.background}
            isEstimate={summary.isEstimate}
            className="text-white"
          />
        </tbody>
      </table>

      {children ? (
        <div className="flex justify-center px-1 mt-6 mb-4">{children}</div>
      ) : null}
    </div>
  );
}
