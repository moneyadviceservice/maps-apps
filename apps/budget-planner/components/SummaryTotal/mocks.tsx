import NumberFormat from '@maps-react/common/components/NumberFormat';

export const background = 'bg-green-700';
interface Item {
  label: string;
  value: JSX.Element;
  isEstimate: boolean;
}

const createItem = (
  label: string,
  value: number,
  direction?: 'increase' | 'decrease',
  isEstimate = false,
): Item => ({
  label,
  value: (
    <NumberFormat
      prefix="Â£"
      renderText={(value) => <b>{value}</b>}
      value={value}
      direction={direction}
    />
  ),
  isEstimate,
});

export const items: Item[] = [
  createItem('Income', 1000),
  createItem('Spending', 200),
];

export const itemsNeutral: Item[] = [
  createItem('Income', 1000),
  createItem('Spending', 1000),
];

export const itemsNegative: Item[] = [
  createItem('Income', 1000),
  createItem('Spending', Math.abs(-1200)),
];

export const itemsIncrease: Item[] = [
  createItem('Your income', 1400),
  createItem('Your budget now', 1100),
  createItem('Your new budget', 1100, 'increase'),
];

export const itemsDecrease: Item[] = [
  createItem('Your income', 1400),
  createItem('Your budget now', 1100),
  createItem('Your new budget', 980, 'decrease'),
];

export const itemsWithEstimate: Item[] = [
  createItem('Estimated Income', 1000, undefined, true),
  createItem('Estimated Spending', 200, undefined, true),
];
