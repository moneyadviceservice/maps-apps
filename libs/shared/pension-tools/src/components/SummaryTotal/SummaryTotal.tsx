import { twMerge } from 'tailwind-merge';

import { H6 } from '@maps-react/common/components/Heading';
import NumberFormat from '@maps-react/common/components/NumberFormat';
import { Paragraph } from '@maps-react/common/components/Paragraph';

export enum SUMMARY_TOTAL_STATUS_TYPES {
  POSITIVE = 'positive',
  NEGATIVE = 'negative',
  BALANCED = 'balanced',
}

type Props = {
  title: string;
  income: number;
  spending: number;
  balance: number;
  incomeLabel: string;
  spendingLabel: string;
  balanceLabel: string;
  status?: SUMMARY_TOTAL_STATUS_TYPES;
  'data-testid'?: string;
};

export const SummaryTotal = ({
  'data-testid': testId,
  title,
  income,
  incomeLabel,
  spending,
  spendingLabel,
  balance,
  balanceLabel,
  status = SUMMARY_TOTAL_STATUS_TYPES.BALANCED,
}: Props) => {
  const mainClasses = [
    'py-[22px]',
    'pl-4',
    'pr-6',
    'rounded',
    'text-lg',
    'grid',
    'grid-cols-3',
  ];
  return (
    <div
      data-testid={testId}
      className="border border-slate-400 rounded px-2 pt-6 pb-4 h-fit space-y-2"
    >
      <H6 className="text-xl pl-2 pb-2">{title}</H6>
      {[
        { label: incomeLabel, value: income },
        { label: spendingLabel, value: spending },
      ].map((item, idx) => (
        <div key={idx} className={twMerge(mainClasses, 'bg-gray-100')}>
          <Paragraph className="col-span-2 m-0">{item.label}</Paragraph>
          <NumberFormat
            className="justify-self-end"
            prefix="£"
            value={item.value}
          />
        </div>
      ))}

      <div
        className={twMerge(
          mainClasses,
          'text-white',
          status === SUMMARY_TOTAL_STATUS_TYPES.BALANCED && 'bg-gray-650',
          status === SUMMARY_TOTAL_STATUS_TYPES.POSITIVE && 'bg-green-700',
          status === SUMMARY_TOTAL_STATUS_TYPES.NEGATIVE && 'bg-red-700',
        )}
      >
        <Paragraph className="col-span-2 m-0">{balanceLabel}</Paragraph>
        <NumberFormat className="justify-self-end" prefix="£" value={balance} />
      </div>
    </div>
  );
};
