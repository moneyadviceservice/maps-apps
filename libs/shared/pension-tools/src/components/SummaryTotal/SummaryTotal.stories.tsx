import { StoryFn } from '@storybook/react';

import SummaryTotal, { SUMMARY_TOTAL_STATUS_TYPES } from './SummaryTotal';

const StoryProps = {
  title: 'Components/PENSION-TOOLS/SummaryTotal',
  component: SummaryTotal,
  argTypes: {},
};

const Template: StoryFn<typeof SummaryTotal> = (args) => (
  <SummaryTotal {...args} />
);

export default StoryProps;
export const Default = Template.bind({});
const baseargs = {
  title: 'Summary total',
  income: 2000,
  spending: 1500,
  balance: 500,
  incomeLabel: 'Income',
  spendingLabel: 'Spending',
  balanceLabel: 'Balance',
};
Default.args = {
  ...baseargs,
  status: SUMMARY_TOTAL_STATUS_TYPES.BALANCED,
};

export const Positive = Template.bind({});
Positive.args = {
  ...baseargs,
  status: SUMMARY_TOTAL_STATUS_TYPES.POSITIVE,
};

export const Negative = Template.bind({});
Negative.args = {
  ...baseargs,
  status: SUMMARY_TOTAL_STATUS_TYPES.NEGATIVE,
};
