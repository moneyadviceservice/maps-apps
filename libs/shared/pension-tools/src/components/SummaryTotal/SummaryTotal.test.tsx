import { render } from '@testing-library/react';

import { SUMMARY_TOTAL_STATUS_TYPES, SummaryTotal } from './SummaryTotal';

const mockParams = {
  title: 'Summary total',
  income: 4000,
  spending: 3000,
  balance: 1000,
  incomeLabel: 'Income',
  spendingLabel: 'Spending',
  balanceLabel: 'Balance',
};

describe('Summary Total component', () => {
  it('should render Summary Total component in balanced state', () => {
    const { container } = render(<SummaryTotal {...mockParams} />);

    expect(container).toMatchSnapshot();
  });

  it('should render Summary Total component in positive state', () => {
    const { container } = render(
      <SummaryTotal
        {...mockParams}
        status={SUMMARY_TOTAL_STATUS_TYPES.POSITIVE}
      />,
    );

    expect(container).toMatchSnapshot();
  });

  it('should render Summary Total component in negative state', () => {
    const { container } = render(
      <SummaryTotal
        {...mockParams}
        status={SUMMARY_TOTAL_STATUS_TYPES.NEGATIVE}
      />,
    );

    expect(container).toMatchSnapshot();
  });
});
