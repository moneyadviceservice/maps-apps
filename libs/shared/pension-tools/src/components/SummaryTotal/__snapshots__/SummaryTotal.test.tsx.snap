// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Summary Total component should render Summary Total component in balanced state 1`] = `
<div>
  <div
    class="border border-slate-400 rounded px-2 pt-6 pb-4 h-fit space-y-2"
  >
    <h6
      class="text-gray-800 font-bold text-xl pl-2 pb-2"
    >
      Summary total
    </h6>
    <div
      class="py-[22px] pl-4 pr-6 rounded text-lg grid grid-cols-3 bg-gray-100"
    >
      <p
        class="col-span-2 m-0"
        data-testid="paragraph"
      >
        Income
      </p>
      <span
        class="justify-self-end"
      >
        £4,000
      </span>
    </div>
    <div
      class="py-[22px] pl-4 pr-6 rounded text-lg grid grid-cols-3 bg-gray-100"
    >
      <p
        class="col-span-2 m-0"
        data-testid="paragraph"
      >
        Spending
      </p>
      <span
        class="justify-self-end"
      >
        £3,000
      </span>
    </div>
    <div
      class="py-[22px] pl-4 pr-6 rounded text-lg grid grid-cols-3 text-white bg-gray-650"
    >
      <p
        class="col-span-2 m-0"
        data-testid="paragraph"
      >
        Balance
      </p>
      <span
        class="justify-self-end"
      >
        £1,000
      </span>
    </div>
  </div>
</div>
`;

exports[`Summary Total component should render Summary Total component in negative state 1`] = `
<div>
  <div
    class="border border-slate-400 rounded px-2 pt-6 pb-4 h-fit space-y-2"
  >
    <h6
      class="text-gray-800 font-bold text-xl pl-2 pb-2"
    >
      Summary total
    </h6>
    <div
      class="py-[22px] pl-4 pr-6 rounded text-lg grid grid-cols-3 bg-gray-100"
    >
      <p
        class="col-span-2 m-0"
        data-testid="paragraph"
      >
        Income
      </p>
      <span
        class="justify-self-end"
      >
        £4,000
      </span>
    </div>
    <div
      class="py-[22px] pl-4 pr-6 rounded text-lg grid grid-cols-3 bg-gray-100"
    >
      <p
        class="col-span-2 m-0"
        data-testid="paragraph"
      >
        Spending
      </p>
      <span
        class="justify-self-end"
      >
        £3,000
      </span>
    </div>
    <div
      class="py-[22px] pl-4 pr-6 rounded text-lg grid grid-cols-3 text-white bg-red-700"
    >
      <p
        class="col-span-2 m-0"
        data-testid="paragraph"
      >
        Balance
      </p>
      <span
        class="justify-self-end"
      >
        £1,000
      </span>
    </div>
  </div>
</div>
`;

exports[`Summary Total component should render Summary Total component in positive state 1`] = `
<div>
  <div
    class="border border-slate-400 rounded px-2 pt-6 pb-4 h-fit space-y-2"
  >
    <h6
      class="text-gray-800 font-bold text-xl pl-2 pb-2"
    >
      Summary total
    </h6>
    <div
      class="py-[22px] pl-4 pr-6 rounded text-lg grid grid-cols-3 bg-gray-100"
    >
      <p
        class="col-span-2 m-0"
        data-testid="paragraph"
      >
        Income
      </p>
      <span
        class="justify-self-end"
      >
        £4,000
      </span>
    </div>
    <div
      class="py-[22px] pl-4 pr-6 rounded text-lg grid grid-cols-3 bg-gray-100"
    >
      <p
        class="col-span-2 m-0"
        data-testid="paragraph"
      >
        Spending
      </p>
      <span
        class="justify-self-end"
      >
        £3,000
      </span>
    </div>
    <div
      class="py-[22px] pl-4 pr-6 rounded text-lg grid grid-cols-3 text-white bg-green-700"
    >
      <p
        class="col-span-2 m-0"
        data-testid="paragraph"
      >
        Balance
      </p>
      <span
        class="justify-self-end"
      >
        £1,000
      </span>
    </div>
  </div>
</div>
`;
